README
[![Code Climate](https://codeclimate.com/github/malachaifrazier/deadlock_retry/badges/gpa.svg)](https://codeclimate.com/github/malachaifrazier/deadlock_retry)
[![Test Coverage](https://codeclimate.com/github/malachaifrazier/deadlock_retry/badges/coverage.svg)](https://codeclimate.com/github/malachaifrazier/deadlock_retry)
[![Coverage Status](https://coveralls.io/repos/malachaifrazier/deadlock_retry/badge.png)](https://coveralls.io/r/malachaifrazier/deadlock_retry)
[![Dependency Status](https://gemnasium.com/malachaifrazier/deadlock_retry.svg)](https://gemnasium.com/malachaifrazier/deadlock_retry)


= Deadlock Retry

Deadlock retry allows the database adapter (currently only tested with the
MySQLAdapter) to retry transactions that fall into deadlock. It will retry
such transactions three times before finally failing.

This capability is automatically added to ActiveRecord. No code changes or otherwise are required.

== Installation

Add it to your Rails application by installing the gem:

  gem install deadlock_retry

and including a reference to it in your application's Gemfile:

  gem 'deadlock_retry'


CHANGELOG
deadlock_retry changes

== v1.2.1 - (10/16/2014)

* .travis.yml added
* codeclimate added

== v1.2.0

* Support for postgres (tomhughes)
* Testing AR versions (kbrock)

== v1.1.2

* Exponential backoff, sleep 0, 1, 2, 4... seconds between retries.
* Support new syntax for InnoDB status in MySQL 5.5.

== v1.1.1 (2011-05-13)

* Conditionally log INNODB STATUS only if user has permission. (osheroff)

== v1.1.0 (2011-04-20)

* Modernize.
* Drop support for Rails 2.1 and earlier.

== v1.0 - (2009-02-07)

* Add INNODB status logging for debugging deadlock issues.
* Clean up so the code will run as a gem plugin.
* Small fix for ActiveRecord 2.1.x compatibility.


test/test_helper.rb
$:.unshift(File.join(File.dirname(__FILE__),'..','lib'))
# $:.push File.expand_path("../lib", __FILE__)
require 'codeclimate-test-reporter'
require 'rubygems'
require 'active_record'
require 'active_record/version'
puts "Testing ActiveRecord #{ActiveRecord::VERSION::STRING}"
require 'minitest/autorun'
require 'minitest/spec'
require 'mocha/setup'
require 'mocha/mini_test'
require 'logger'
# require '../lib/deadlock_retry'
require 'coveralls'
gem 'activerecord', ENV['ACTIVERECORD_VERSION'] || ' ~>3.2'
Coveralls.wear!
CodeClimate::TestReporter.start



test/deadlock_retry_test.rb
require 'rubygems'

# # Change the version if you want to test a different version of ActiveRecord
gem 'activerecord', ENV['ACTIVERECORD_VERSION'] || ' ~>3.2'
require 'active_record'
require 'active_record/version'
puts "Testing ActiveRecord #{ActiveRecord::VERSION::STRING}"

require 'minitest/autorun'
require 'mocha/mini_test'
require 'logger'
# require '../lib/deadlock_retry'
require 'test_helper.rb'

class MockModel
  # include DeadlockRetry
  @@open_transactions = 0

  def self.transaction(*objects)
    @@open_transactions += 1
    yield
  ensure
    @@open_transactions -= 1
  end

  def self.open_transactions
    @@open_transactions
  end

  def self.connection
    self
  end

  def self.logger
    @logger ||= Logger.new(nil)
  end

  def self.show_innodb_status
    []
  end

  def self.select_rows(sql)
    [['version', '5.1.45']]
  end

  def self.select_value(sql)
    true
  end

  def self.adapter_name
    "MySQL"
  end
end

class DeadlockRetryTest < Minitest::Unit::TestCase
  DEADLOCK_ERROR = "MySQL::Error: Deadlock found when trying to get lock"
  TIMEOUT_ERROR = "MySQL::Error: Lock wait timeout exceeded"
  # include DeadlockRetry

  def setup
    MockModel.stubs(:exponential_pause)
  end

  def test_no_errors
    assert_equal :success, MockModel.transaction { :success }
  end

  def test_no_errors_with_deadlock
    errors = [ DEADLOCK_ERROR ] * 3
    binding.pry
    assert_equal :success, MockModel.transaction { raise ActiveRecord::StatementInvalid, errors.shift unless errors.empty?; :success }
    assert errors.empty?
  end

  def test_no_errors_with_lock_timeout
    errors = [ TIMEOUT_ERROR ] * 3
    assert_equal :success, MockModel.transaction { raise ActiveRecord::StatementInvalid, errors.shift unless errors.empty?; :success }
    assert errors.empty?
  end

  def test_error_if_limit_exceeded
    assert_raises(ActiveRecord::StatementInvalid) do
      MockModel.transaction { raise ActiveRecord::StatementInvalid, DEADLOCK_ERROR }
    end
  end

  def test_error_if_unrecognized_error
    assert_raises(ActiveRecord::StatementInvalid) do
      MockModel.transaction { raise ActiveRecord::StatementInvalid, "Something else" }
    end
  end

  def test_included_by_default
    assert ActiveRecord::Base.respond_to?(:transaction_with_deadlock_handling)
  end

  def test_innodb_status_availability
    DeadlockRetry.innodb_status_cmd = nil
    MockModel.transaction {}
    assert_equal "show innodb status", DeadlockRetry.innodb_status_cmd
  end


  def test_error_in_nested_transaction_should_retry_outermost_transaction
    tries = 0
    errors = 0

    MockModel.transaction do
      tries += 1
      MockModel.transaction do
        MockModel.transaction do
          errors += 1
          raise ActiveRecord::StatementInvalid, "MySQL::Error: Lock wait timeout exceeded" unless errors > 3
        end
      end
    end

    assert_equal 4, tries
  end
end


lib/deadlock_retry.rb
require 'active_support/core_ext/module/attribute_accessors'

module DeadlockRetry
  def self.included(base)
    base.extend(ClassMethods)
    base.class_eval do
      class << self
        alias_method_chain :transaction, :deadlock_handling
      end
    end
  end

  mattr_accessor :innodb_status_cmd

  module ClassMethods
    DEADLOCK_ERROR_MESSAGES = [
      "Deadlock found when trying to get lock",
      "Lock wait timeout exceeded",
      "deadlock detected"
    ]

    MAXIMUM_RETRIES_ON_DEADLOCK = 5


    def transaction_with_deadlock_handling(*objects, &block)
      retry_count = 0

      check_innodb_status_available

      begin
        transaction_without_deadlock_handling(*objects, &block)
      rescue ActiveRecord::StatementInvalid => error
        raise if in_nested_transaction?
        if DEADLOCK_ERROR_MESSAGES.any? { |msg| error.message =~ /#{Regexp.escape(msg)}/ }
          raise if retry_count >= MAXIMUM_RETRIES_ON_DEADLOCK
          retry_count += 1
          logger.info "Deadlock detected on retry #{retry_count}, restarting transaction"
          log_innodb_status if DeadlockRetry.innodb_status_cmd
          exponential_pause(retry_count)
          retry
        else
          raise
        end
      end
    end

    private

    WAIT_TIMES = [0, 1, 2, 4, 8, 16, 32]

    def exponential_pause(count)
      sec = WAIT_TIMES[count-1] || 32
      # sleep 0, 1, 2, 4, ... seconds up to the MAXIMUM_RETRIES.
      # Cap the pause time at 32 seconds.
      sleep(sec) if sec != 0
    end

    def in_nested_transaction?
      # open_transactions was added in 2.2's connection pooling changes.
      connection.open_transactions != 0
    end

    def show_innodb_status
       self.connection.select_value(DeadlockRetry.innodb_status_cmd)
    end

    # Should we try to log innodb status -- if we don't have permission to,
    # we actually break in-flight transactions, silently (!)
    def check_innodb_status_available
      return unless DeadlockRetry.innodb_status_cmd == nil

      if self.connection.adapter_name == "MySQL"
        begin
          mysql_version = self.connection.select_rows('show variables like \'version\'')[0][1]
          cmd = if mysql_version < '5.5'
            'show innodb status'
          else
            'show engine innodb status'
          end
          self.connection.select_value(cmd)
          DeadlockRetry.innodb_status_cmd = cmd
        rescue
          logger.info "Cannot log innodb status: #{$!.message}"
          DeadlockRetry.innodb_status_cmd = false
        end
      else
        DeadlockRetry.innodb_status_cmd = false
      end
    end

    def log_innodb_status
      # show innodb status is the only way to get visiblity into why
      # the transaction deadlocked.  log it.
      lines = show_innodb_status
      logger.warn "INNODB Status follows:"
      lines.each_line do |line|
        logger.warn line
      end
    rescue => e
      # Access denied, ignore
      logger.info "Cannot log innodb status: #{e.message}"
    end

  end
end

ActiveRecord::Base.send(:include, DeadlockRetry) if defined?(ActiveRecord)
